src/dinit-iostream.cc

```diff
@@ -0,0 +1,553 @@
+#include <algorithm>
+#include <exception>
+#include <limits>
+#include <cstring>
+
+#include <dinit-iostream.h>
+#include <baseproc-sys.h>
+
+// See dinit-iostream.h for inforamtion about the usage
+
+namespace {
+
+// A simple trick for calculating how much digits can be in a type with compile-time
+// integer logarithm like function.
+template <typename T> constexpr unsigned TYPE_MAX_NUMDIGITS(T num = std::numeric_limits<T>::max(),
+        unsigned pow = 0)
+{
+    return (num == 0) ? pow : TYPE_MAX_NUMDIGITS(num / 10, pow + 1);
+}
+
+constexpr unsigned INT_NUMDIGITS = TYPE_MAX_NUMDIGITS<int>() + 1;
+constexpr unsigned UINT_NUMDIGITS = TYPE_MAX_NUMDIGITS<unsigned>() + 1;
+constexpr unsigned LONG_NUMDIGITS = TYPE_MAX_NUMDIGITS<long>() + 1;
+constexpr unsigned ULONG_NUMDIGITS = TYPE_MAX_NUMDIGITS<unsigned long>() + 1;
+
+} /* anonymous namespace */
+
+namespace dio {
+
+// class streambuf_base
+streambuf *streambuf_base::get_buf() noexcept
+{
+    return buf.get();
+}
+
+// class ostream
+ssize_t ostream::put(const char *msg, size_t count) noexcept
+{
+    if (!buf) {
+        buffer_failbit = true;
+        return -1; // Tried to use output on null pointer buffer
+    }
+
+    // Capturing flush_nothrow() calls for returing exat number of written characters
+    size_t written_characters = 0;
+
+    if (msg == nullptr || count == 0) {
+        return 0; // Null/Empty message
+    }
+    else if (count > buf->get_size()) {
+        ssize_t r = flush_nothrow();
+        if (r < 0) {
+            // posixio_errno was set by flush_nothrow() call
+            return -1;
+        }
+        ssize_t res = bp_sys::write(fd, msg, count);
+        if (res < 0) {
+            posixio_errno = errno;
+            return -1;
+        }
+        return (r > 0) ? r + res : -1;
+    }
+    else while (count > buf->get_free()) {
+        // If we haven't enough storage for caputring the message
+        // Firstly we try to free-up the buffer and then continue the
```

Comment:
The comment doesn't quite match the code. First it fills up the buffer, _then_ it flushes the buffer. Also `posixio_errno` isn't set here, that happens in `flush_nothrow()`.